packages/intrepid2/src/Discretization/Integration/Intrepid2_IntegrationToolsDef.hpp

```diff
@@ -0,0 +1,2322 @@
+// @HEADER
+// ************************************************************************
+//
+//                           Intrepid2 Package
+//                 Copyright (2007) Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_IntegrationToolsDef.hpp
+    \brief  Definition file for the Intrepid2::IntegrationTools class.
+    \author Created by Nathan V. Roberts.
+*/
+
+#ifndef __INTREPID2_INTEGRATIONTOOLS_DEF_HPP__
+#define __INTREPID2_INTEGRATIONTOOLS_DEF_HPP__
+
+#include "Intrepid2_FunctorIterator.hpp"
+#include "Intrepid2_TensorArgumentIterator.hpp"
+
+#include "Teuchos_TimeMonitor.hpp"
+
+namespace Intrepid2 {
+
+  namespace Impl
+  {
+    /**
+      \brief Implementation of a general sum factorization algorithm, abstracted from the algorithm described by Mora and Demkowicz, for integration.  Uses hierarchical parallelism.
```

Comment:
Unfortunately, not as such/not yet.  All the code here has been unit tested at one time or another, but so far the choice of which integration algorithm to use is based on hard-coded choices in `integrate()`.  There is a TODO within the integrate() implementation to rectify this:
```
// TODO: expose the options for forceNonSpecialized and usePointCacheForRank3Tensor through an IntegrationAlgorithm enumeration.
// AUTOMATIC: let Intrepid2 choose an algorithm based on the inputs (and, perhaps, the execution space)
// STANDARD: don't use sum factorization or axis alignment -- just do the simple contraction, a (p+1)^9 algorithm in 3D
// SUM_FACTORIZATION                              // (p+1)^7 algorithm in 3D
// SUM_FACTORIZATION_AXIS_ALIGNED                 // (p+1)^6 algorithm in 3D
// SUM_FACTORIZATION_FORCE_GENERIC_IMPLEMENTATION // mainly intended for testing purposes (specialized implementations perform better when they are provided)
// SUM_FACTORIZATION_WITH_POINT_CACHE             // novel (p+1)^7 (in 3D) algorithm in Intrepid2; unclear as yet when and whether this may be a superior approach
```

Once this is implemented, we can exercise each of the implementations through unit tests; doing so would actually not be too much effort, given the tests we already have.  But I'd like to get this PR in first.